### PR TITLE
Issue 2613 custom template GitHub

### PR DESCRIPTION
**Pull request by ShubhamRasal**
## Proposed changes
Add custom template download/update support from github
    
    - Accept the -gtr flag to accept the list of custom template
      repos(public/private)
    - Accept the -gt flag for github token. It internally sets os.Env
      variable
    - Update the flags from
       - -update to -nuclei-update for nuclei self update
       - -ut to -tup for template-update
       - -ud to -tud for custom template location
    - Add github.go file which has code related to download and update
      custom templates repos.

## Checklist

- [x] Pull request is created against the [dev](https://github.com/projectdiscovery/nuclei/tree/dev) branch
- [x] All checks passed (lint, unit/integration/regression tests etc.) with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Review by ehsandeep — CHANGES_REQUESTED** [review]
- [ ] Adding support for `GITHUB_TEMPLATE_REPO` env variable
- [ ] Default to the existing directory structure instead of `community` directory.
- [ ] `github` directory for a custom GitHub template in the current default location.

File: v2/internal/runner/github.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+	"github.com/google/go-github/github"
+	"github.com/pkg/errors"
+	"gopkg.in/src-d/go-git.v4/plumbing/transport/http"
+)
+
+// This function download the custom template repositories in given location
+// scenario 1: -gtr custom-template.txt  flag has passed => Only download the repos. Do not update
+// scenario 2: -gtr custom-template.txt -tup github   => Update the repo(git pull)  and download if any new repo
+// Reason to add update and download logic in single function is scenario 2
+func (r *Runner) downloadCustomTemplateRepo(repoName string, ctx context.Context) (string, error) {
```
Comment (jdk2588) [inline v2/internal/runner/github.go:20]: Where is the Runner coming from? It seems to be in some different file.

Comment (jdk2588) [reply v2/internal/runner/github.go:20]: This should be part of runner.go.

**Review by jdk2588 — CHANGES_REQUESTED** [review]

File: v2/internal/runner/github.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+	for _, customTemplate := range r.customTemplates {
+		clonePath := customTemplate.getLocalRepoClonePath(downloadPath)
+		if r.options.UpdateTemplates == "github" && isDirectory(clonePath) {
+			err := customTemplate.pullChanges(clonePath, r.options.GithubToken)
+			if err == nil {
+				gologger.Info().Label("GITHUB").Msgf("Repo %s/%s successfully pulled the changes.\n", customTemplate.owner, customTemplate.reponame)
+			} else {
+				gologger.Error().Label("GITHUB").Msgf("%s", err)
+			}
+		} else {
```
Comment (jdk2588) [inline v2/internal/runner/github.go:37]: Use guard code with if condition.

**Review by jdk2588 — CHANGES_REQUESTED** [review]

File: v2/internal/runner/github.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+// scenario 1: -gtr custom-template.txt  flag has passed => Only download the repos. Do not update
+// scenario 2: -gtr custom-template.txt -tup github   => Update the repo(git pull)  and download if any new repo
+// Reason to add update and download logic in single function is scenario 2
+func (r *Runner) downloadCustomTemplates(ctx context.Context) {
+	downloadPath := filepath.Join(r.templatesConfig.TemplatesDirectory, customTemplateType)
+
+	for _, customTemplate := range r.customTemplates {
+		clonePath := customTemplate.getLocalRepoClonePath(downloadPath)
+
+		if !isDirectory(clonePath) {
```
Comment (Mzack9999) [inline v2/internal/runner/github.go:32]: what about `fileutil.FolderExists`

File: v2/internal/runner/github.go [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+			if err != nil {
+				gologger.Info().Msgf("%s", err)
+			} else {
+				gologger.Info().Msgf("Repo %s/%s cloned successfully at %s", customTemplate.owner, customTemplate.reponame, clonePath)
+			}
+			continue
+		}
+		if r.options.UpdateTemplates == customTemplateType {
+			err := customTemplate.pullChanges(clonePath, r.options.GithubToken)
+			if err == nil {
```
Comment (Mzack9999) [inline v2/internal/runner/github.go:43]: really just a minor detail: I'd suggest handling `err != nil` first

File: v2/internal/runner/github.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+// getOwnerAndRepo returns the owner, repo, err from the given string
+// eg. it takes input projectdiscovery/nuclei-templates and
+// returns owner=> projectdiscovery , repo => nuclei-templates
+func getOwnerAndRepo(reponame string) (owner string, repo string, err error) {
+	s := strings.Split(reponame, "/")
+	if len(s) != 2 {
+		err = errors.Errorf("wrong Repo name: %s", reponame)
+		return
+	}
+	owner = strings.Split(reponame, "/")[0]
```
Comment (Mzack9999) [inline v2/internal/runner/github.go:86]: we can reuse the variable as we already performed the split:
`strings.Split(reponame, "/")` => `s`

File: v2/internal/runner/github.go [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+	repo = strings.Split(reponame, "/")[1]
```
Comment (Mzack9999) [inline v2/internal/runner/github.go:87]: we can reuse the variable as we already performed the split:
`strings.Split(reponame, "/")` => `s`

File: v2/internal/runner/github.go [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+		return errors.Errorf("%s/%s: %s", ctr.owner, ctr.reponame, err.Error())
+	}
+	return nil
+}
+
+// getLocalRepoClonePath returns the clone path.
+// if same name repo directory exists from another owner then it appends the owner then and returns the path
+// eg. for nuclei-templates directory exists for projectdiscovery owner, then for ehsandeep/nuclei-templates it will return nuclei-templates-ehsandeep
+func (ctr *customTemplateRepo) getLocalRepoClonePath(downloadPath string) string {
+	if isDirectory(filepath.Join(downloadPath, ctr.reponame)) && !ctr.isRepoDirExists(filepath.Join(downloadPath, ctr.reponame)) {
```
Comment (Mzack9999) [inline v2/internal/runner/github.go:146]: consider using `fileutil.FolderExists`

File: v2/internal/runner/github.go [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+// returns the auth object with username and github token as password
+func getAuth(username, password string) *http.BasicAuth {
+	if username != "" && password != "" {
+		return &http.BasicAuth{Username: username, Password: password}
+	}
+	return nil
+}
+
+// isDirectory returns true if path is directory
+func isDirectory(path string) bool {
```
Comment (Mzack9999) [inline v2/internal/runner/github.go:168]: please check if `fileutil.FolderExists` already covers this check

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
suggesting minor optimizations

File: v2/internal/runner/github.go (hunk 8 above)
Comment (ShubhamRasal) [reply v2/internal/runner/github.go]: Updated as per your suggestion

File: v2/internal/runner/github.go (hunk 7 above)
Comment (ShubhamRasal) [reply v2/internal/runner/github.go]: Updated as per your suggestion

File: v2/internal/runner/github.go (hunk 3 above)
Comment (ShubhamRasal) [reply v2/internal/runner/github.go]: Updated as per your suggestion

**Review by Mzack9999 — APPROVED** [review]
Note: contains breaking change due to update flag renaming

**Review by ehsandeep — CHANGES_REQUESTED** [review]
@ShubhamRasal, we can update `-update-template` option to keep it the same as before, i.e boolean flag, disabled as default, and if used, it will check for update for both options community / GitHub implicitly.

```console
   -ut, -update-template          update nuclei-templates to latest released version
```

**Review by ehsandeep — COMMENTED** [review]
@ShubhamRasal, this is working well, we can also update the [nuclei-docs](https://github.com/projectdiscovery/nuclei-docs) mentioning about the use of custom / github template as per this change in the [template](https://github.com/projectdiscovery/nuclei-docs/blob/master/docs/nuclei/get-started.md#nuclei-templates) section.

**Review by Ice3man543 — APPROVED** [review]

**Pull request merged**